File: Recent_Search/v1/quick-starts/search-quickstart.rb
require "net/https"
require "uri"
require 'base64'
require 'json'

#Quote your Twitter App consumer keys here.
CONSUMER_KEY = "" # Add your API key here
CONSUMER_SECRET = "" # Add your API secret key here

@search_url = "https://api.twitter.com/labs/1/tweets/search"
query = '(Labs Search Twitter) OR from:TwitterDev OR from:SnowBotDev OR from:DailyNASA'

options = {"query" => query, "format" => 'compact'}

def bearer_token(consumer_key, consumer_secret)
# Generates a Bearer Token using your Twitter App's consumer key and secret.
# Calls the Twitter URL below and returns the Bearer Token.
  bearer_token_url = "https://api.twitter.com/oauth2/token"

  return @bearer_token unless @bearer_token.nil?

  credentials = Base64.encode64("#{consumer_key}:#{consumer_secret}").gsub("\n", "")

  uri = URI(bearer_token_url)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  request = Net::HTTP::Post.new(uri.path)
  request.body =  "grant_type=client_credentials"
  request['Authorization'] = "Basic #{credentials}"
  request['User-Agent'] = "LabsRecentSearchQuickStartRuby"

  response = http.request(request)

  body = JSON.parse(response.body)

  body['access_token']

end

def make_request(key, secret, query)

  uri = URI(@search_url)

  options = {}
  options['query'] = query
  options['max_results'] = 10
  options['format'] = 'compact'
  #options['start_time'] = nil
  #options['end_time'] = nil

  uri.query = URI.encode_www_form(options)

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  request = Net::HTTP::Get.new(uri.request_uri)
  request['Authorization'] = "Bearer #{bearer_token(key,secret)}"
  request['User-Agent'] = "RecentSearchQuickStartRuby"

  response = http.request(request)

  return response
end

response = make_request(CONSUMER_KEY, CONSUMER_SECRET, query)

puts JSON.pretty_generate(JSON.parse(response.body))

File: Recent_Search/v1/quick-starts/search-quickstart.py
import json
import urllib.parse
import requests
from requests.auth import AuthBase

# Fill these in. Generate tokens at https://developer.twitter.com/en/apps. 
CONSUMER_KEY = ''
CONSUMER_SECRET = ''

query = urllib.parse.quote("(Labs Search Twitter) OR from:TwitterDev OR from:SnowBotDev OR from:DailyNASA")

url = f"https://api.twitter.com/labs/1/tweets/search?query={query}"

headers = {
    "Accept-Encoding": "gzip"
}

# Generates a bearer token with consumer key and secret via https://api.twitter.com/oauth2/token.
class BearerTokenAuth(AuthBase):
    def __init__(self, consumer_key, consumer_secret):
        self.bearer_token_url = "https://api.twitter.com/oauth2/token"
        self.consumer_key = consumer_key
        self.consumer_secret = consumer_secret
        self.bearer_token = self.get_bearer_token()

    def get_bearer_token(self):
        response = requests.post(
            self.bearer_token_url,
            auth=(self.consumer_key, self.consumer_secret),
            data={'grant_type': 'client_credentials'},
            headers={'User-Agent': 'LabsRecentSearchQuickStartPython'})

        if response.status_code is not 200:
            raise Exception("Cannot get a Bearer token (HTTP %d): %s" % (response.status_code, response.text))

        body = response.json()
        return body['access_token']

    def __call__(self, r):
        r.headers['Authorization'] = f"Bearer %s" % self.bearer_token
        r.headers['User-Agent'] = 'LabsRecentSearchQuickStartPython'
        return r

# Script starts here.

#Create Bearer Token for authenticating with recent search.
bearer_token = BearerTokenAuth(CONSUMER_KEY, CONSUMER_SECRET)

#Make a GET request to the Labs recent search endpoint.
response = requests.get(url, auth=bearer_token, headers = headers)

if response.status_code is not 200:
    raise Exception(f"Request reurned an error: %s" % (response.status_code, response.text))

#Display the returned Tweet JSON.
parsed = json.loads(response.text)
pretty_print = json.dumps(parsed, indent=2, sort_keys=True)
print (pretty_print)
